File: 2017/03/spiral-memory.go
package main

import "fmt"
import "os"
import "math"
import "strconv"


func calculateSpiralManhattan(square int) int {
    if square == 1 {
        return 0
    }

    // Find the first axis.
    squareFloor := int(math.Sqrt(float64(square - 1)))
    ringMaxRoot := squareFloor + (squareFloor % 2) + 1
    ringMax := int(math.Pow(float64(ringMaxRoot), 2))
    ringId := (ringMaxRoot + 1) / 2 - 1

    // Find the perpendicular axis.
    ringSize := ringMax - int(math.Pow(math.Sqrt(float64(ringMax)) - 2, 2))
    edgeIndex := ringSize - (ringMax - square)
    perpendicular := int(math.Abs(float64(edgeIndex % (ringId * 2) - ringId)))

    // Manhattan distance is sum of axis movements
    distance := ringId + perpendicular
    return distance
}


func readSquare(square string) int {
    n, err := strconv.Atoi(square)
    if err != nil {
        panic(err)
    }
    return n
}


func main() {
    var square = readSquare(os.Args[1])
    var distance = calculateSpiralManhattan(square)
    fmt.Println(distance)
}

File: 2017/03/spiral-memory.py
# -*- coding: utf-8 -*-
import math
import sys


def calculate_spiral_manhattan(square):
    if square == 1:
        return 0

    square_floor = int(math.sqrt(square - 1))
    ring_max_root = square_floor + (square_floor % 2) + 1
    ring_max = ring_max_root ** 2
    ring_id = (ring_max_root + 1) // 2 - 1

    ring_size = ring_max - (math.sqrt(ring_max) - 2) ** 2
    edge_index = ring_size - (ring_max - square)
    perpendicular = abs(edge_index % (ring_id * 2) - ring_id)

    distance = ring_id + perpendicular
    return int(distance)


def main(square):
    distance = calculate_spiral_manhattan(square)
    print(distance)


if __name__ == '__main__':
    main(sys.argv[1])
